source/common/grpc/google_grpc_creds_impl.cc

```diff
@@ -4,13 +4,83 @@
 #include "envoy/grpc/google_grpc_creds.h"
 #include "envoy/registry/registry.h"
 
+#include "common/config/datasource.h"
+
 namespace Envoy {
 namespace Grpc {
 
-/**
- * TODO: Create GoogleGrpcCredentialsFactory for each built-in credential type defined in Google
- * gRPC.
- */
+std::shared_ptr<grpc::ChannelCredentials>
+defaultSslChannelCredentials(const envoy::api::v2::core::GrpcService& grpc_service_config,
+                             bool allow_insecure) {
+  const auto& google_grpc = grpc_service_config.google_grpc();
+  std::shared_ptr<grpc::ChannelCredentials> creds = nullptr;
+  if (allow_insecure) {
+    creds = grpc::InsecureChannelCredentials();
+  } else {
+    creds = grpc::SslCredentials(grpc::SslCredentialsOptions());
+  }
+  if (google_grpc.has_channel_credentials() &&
+      google_grpc.channel_credentials().has_ssl_credentials()) {
+    const auto& ssl_credentials = google_grpc.channel_credentials().ssl_credentials();
+    const grpc::SslCredentialsOptions ssl_credentials_options = {
+        .pem_root_certs = Config::DataSource::read(ssl_credentials.root_certs(), true),
+        .pem_private_key = Config::DataSource::read(ssl_credentials.private_key(), true),
+        .pem_cert_chain = Config::DataSource::read(ssl_credentials.cert_chain(), true),
+    };
+    return grpc::SslCredentials(ssl_credentials_options);
+  }
+  return creds;
+}
+
+std::shared_ptr<grpc::ChannelCredentials>
+defaultChannelCredentials(const envoy::api::v2::core::GrpcService& grpc_service_config,
+                          bool allow_insecure) {
+  std::shared_ptr<grpc::ChannelCredentials> creds =
+      defaultSslChannelCredentials(grpc_service_config, allow_insecure);
```

Comment:
return this if allow_insecure is true? I don't think the default implementation should allow call credentials to be enabled on an insecure channel.

I think grpc may drop the call credentials on an insecure channel anyway: https://github.com/grpc/grpc/issues/8142